frontends/PyCDE/src/pycde/fsm.py

```diff
@@ -246,3 +249,76 @@ def construct(ports):
   fsm_hw_mod.__name__ = fsm_name
   fsm_hw_mod.__module__ = fsm_name
   return fsm_hw_mod
+
+
+def gen_fsm(transitions, name="MyFSM"):
+  """
+  Generate a FSM from a dictionary of states and their transitions.
+
+  E.g.:
+  {
+    "a": [
+      ("c", "go"),
+      "b",
+    ],
+    "b": [],
+    "c": []
+  }
+
+  creates an FSM with 3 states (a, b, c). 'b' and 'c' have no outgoing transitions,
+  and 'a' has two outgoing transitions. The first transition ("c", "go") transitions
+  to 'c' whenever an input port 'go' is asserted. The second transition is a default
+  transition to 'b'.
+
+  Any state and guard referenced within the dictionary will automatically be created
+  as a state operation and top-level input, respectively.
+  """
+
+  class FSM:
+    pass
+
+  # Gather states and input variables
+  states = set()
+  inputs = set()
+  initial = True
+
+  def ensure_state(state, initial=False):
+    if state not in states:
+      setattr(FSM, state, State(initial=initial))
+      states.add(state)
+    return getattr(FSM, state)
+
+  def ensure_input(input):
+    if input not in inputs:
+      setattr(FSM, input, Input(types.i1))
+      inputs.add(input)
+
+  for (state, state_transitions) in transitions.items():
+    currentStateAttr = ensure_state(state, initial)
+    if not type(state_transitions) is list:
+      raise Exception(f"Transitions for state '{state}' must be a list")
+
+    for transition in state_transitions:
+      guard_port = None
+      try:
```

Comment:
Use `if isinstance(...)` instead of this try block. Exceptions are supposed to be for exceptional cases, despite how portions of modern languages/libs are implemented.